nexus/src/app/sagas/demo.rs

```diff
@@ -21,56 +21,66 @@
 use super::NexusActionContext;
 use super::{ActionRegistry, NexusSaga, SagaInitError};
 use crate::app::sagas::declare_saga_actions;
-use anyhow::ensure;
+use anyhow::Context;
 use omicron_common::api::external::Error;
 use omicron_uuid_kinds::DemoSagaUuid;
 use serde::Deserialize;
 use serde::Serialize;
 use slog::info;
 use std::collections::BTreeMap;
+use std::future::Future;
+use std::sync::Arc;
 use steno::ActionError;
-use tokio::sync::oneshot;
+use tokio::sync::Semaphore;
 
-/// Set of demo sagas that have been marked completed
+/// Rendezvous point for demo sagas
 ///
-/// Nexus maintains one of these at the top level.  Individual demo sagas wait
-/// until their id shows up here, then remove it and proceed.
+/// This is where:
+///
+/// - demo sagas wait for a completion message
+/// - completion messages are recorded for demo sagas that haven't started
+///   waiting yet
+///
+/// Nexus maintains one of these structures at the top level.
 pub struct CompletingDemoSagas {
-    ids: BTreeMap<DemoSagaUuid, oneshot::Sender<()>>,
+    sagas: BTreeMap<DemoSagaUuid, Arc<Semaphore>>,
```

Comment:
Could we, perhaps, add some documentation explaining the motivation for using a `Semaphore` with one 0-1 permit, rather than a `Notify`?

Reply:
Ah, I think there is no good reason!  I _thought_ if you called `notify.notify_one()`, but nobody was waiting, then that wakeup would be lost.  I just checked the docs and that's not the case.  Do you think I should change it to use `Notify`?

Reply:
In the interest of fixing the flaky test quickly, I'm going to land anyway, but we can always change this.

Reply:
Yeah, that seems fine to me. I agree that `Notify::notify_one` would also have gotten you the exact same behavior, but it also doesn't actually matter. `Notify` might be _very marginally_ more efficient but this is far from a hot path...